### DISCUSSION TITLE
为什么基础服务不应该高可用 · /whys-the-design-unstable-infrastructure

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#577] [Announcements]
https://draveness.me/whys-the-design-unstable-infrastructure

**Comment by schecterdamien** [comment]
故意触发宕机是不是避免让使用者过分相信基础服务的可用性。这样的话，此时的宕机是自己协议约定内的也是可控的，不用承担过多的责任，但是又能给使用者一个教训，让他们做足准备，在下次发生同样事情的时候减少损失，避免鱼死网破 = =

**Reply by Juminiy** [reply]
参考Netflix的chaosmonkey

**Comment by zcong1993** [comment]
SRE: Google 运维解密 这本书也说了, 故意宕机一方面是让用户明白这些东西不是百分百可靠的, 然后通过自己的手段解决; 另一方面是为了不让用户的预期过高.

**Comment by draveness** [comment]
> SRE: Google 运维解密 这本书也说了, 故意宕机一方面是让用户明白这些东西不是百分百可靠的, 然后通过自己的手段解决; 另一方面是为了不让用户的预期过高.

:+1:

> 故意触发宕机是不是避免让使用者过分相信基础服务的可用性。这样的话，此时的宕机是自己协议约定内的也是可控的，不用承担过多的责任，但是又能给使用者一个教训，让他们做足准备，在下次发生同样事情的时候减少损失，避免鱼死网破 = =

:+1:

**Comment by musele** [comment]
基础服务必须高可用，高可用总的来说就是利用冗余来抵消不确定性，文章标题可能改成《基础服务不应该依赖单一平台的高可用》会好一点。

**Comment by draveness** [comment]
> 基础服务必须高可用，高可用总的来说就是利用冗余来抵消不确定性，文章标题可能改成《基础服务不应该依赖单一平台的高可用》会好一点。

如何实现高可用和应不应该高可用是两件事。

**Comment by yunqingqing** [comment]
为什么我觉得这篇文章讲的是高可用服务也不是100%的高可用

**Comment by jinyuyoulong** [comment]
> @yunqingqing
> 为什么我觉得这篇文章讲的是高可用服务也不是100%的高可用

因为没有完美的100%的高可用，

**Comment by draveness** [comment]
> 为什么我觉得这篇文章讲的是高可用服务也不是100%的高可用

并不是，大多数云服务厂商提供基础服务的可用性都是 3.5 个 9，例如 Redis 等服务，只有极少数的厂商会提供 4 个 9 的可用性，一个月不可用的时间在 4~20min，我并不认为这是高可用，阿里云的云服务器 SLA 如下：

> 2.2  服务可用性承诺
>（1）对于单实例维度， 阿里云承诺一个服务周期内ECS的服务可用性不低于99.95%；
>（2）对于单地域多可用区维度，阿里云承诺一个服务周期内ECS的服务可用性不低于99.99%。

Google 云提供的 Compute Engine Service SLA 是 4 个 9，这在云服务商中已经非常高了

> During the Term of the agreement under which Google has agreed to provide Google Cloud Platform to Customer (as applicable, the "Agreement"), the Covered Service will provide a Monthly Uptime Percentage to Customer of at least 99.99% (the "Service Level Objective" or "SLO")

### Reference 

+ http://terms.aliyun.com/legal-agreement/terms/suit_bu1_ali_cloud/suit_bu1_ali_cloud201803071050_54811.html
+ http://terms.aliyun.com/legal-agreement/terms/suit_bu1_ali_cloud/suit_bu1_ali_cloud201802011632_33742.html
+ https://cloud.google.com/compute/sla

**Comment by draveness** [comment]
> > @yunqingqing
> > 为什么我觉得这篇文章讲的是高可用服务也不是100%的高可用
> 
> 因为没有完美的100%的高可用，

👍 在足够长的时间内，没有服务能做到 100% 的可用，这还是在不考虑收益的情况下。

**Comment by karottc** [comment]
标题党啊。 标题说，不应该做，然而全文都在讲怎么做的尽可能的高可用。

**Comment by draveness** [comment]
> @karottc 标题党啊。 标题说，不应该做，然而全文都在讲怎么做的尽可能的高可用。

你好好读读

**Comment by pangliang** [comment]
业务的高可用不应该依赖基础设施的高可用, 恰恰相反, 业务的高可用应该建立在对基础设施的不稳定的预见性上

**Comment by Liutos** [comment]
故意宕机的做法虽然由于仍然保证了SLA而不需要赔偿，但还挺狠的呀😂

**Comment by draveness** [comment]
> 故意宕机的做法虽然由于仍然保证了SLA而不需要赔偿，但还挺狠的呀😂

人家毕竟是按合同办事，而且如果真的连续几个季度都是 100% 可用，突然变成了 99.95% 可用对我们的影响更大。

**Comment by draveness** [comment]
> 事实上.用基本应用场景相同的两套非常类似的基础组件是不现实的,例如提及的Redis 和 memcached.
> 
> 如果有一套大规模混合存储kv系统(mem+ssd),在此之上套一层纯mem kv系统,这是可以接受的异构
> 
> 用两套不一样的mem kv系统纯粹就是没有大规模企业开发经验而已
>
> 这涉及到日常应用如何最大利用率的使用两套kv存储(redis/memcached),在接口/api/数据结构不一样的情况下,怎么对数据双写,怎么对两个系统做负载均衡,怎么处理数据一致性问题.

这里的场景当然是最优的。但是如果在公司基础层面不支持这种使用方式的情况下，业务方有时间等到公司基础服务支持这些能力么？

这完全取决于公司内部基础框架和服务是否能满足上层业务方的需求，与有没有大规模企业开发经验没有关系。

> 这种场景上,大公司内更容易通过Redis集群(多地多中心容灾)直接提供更高的可用性.避免应用方做些很trick很没有意义的异构实现来保障稳定性.避免上百上千个业务方反复的做同样一套异构系统来保障稳定性.降低整体企业的工程人力消耗.

多地多中心容灾也是一种异构部署的方案，这与文中的内容并不冲突。

感谢这里指出的问题。你讲的很有道理，不过因为我们站在不同的角度去考虑这个问题，最终造成了结果的不同：

+ 从基础服务的角度来看，这种功能最后一定会下沉到基础服务中，由基础服务统一提供能力，这会使公司内部的上层服务都受益；
+ 从业务方的角度来看，除非你的能力足够完善，要不然没有时间等待功能的成熟；

这里忽略的前提是，不是所有的企业都是大企业，也不是所有的团队都有能力自建基础服务，而云服务商提供的可用性往往都是有限的、不够可靠的，这时候应该怎么去提高可用性呢 —— 解决方案就是引入异构，文中没有说异构的服务是最优的选择，只是可以考虑的一个选项。

**Comment by neomerckx** [comment]
就算不是所有企业都是大企业,其实还是不应该在没有特殊需求的情况下混用多种同类型存储,这对系统来说并不能显著提升稳定性,反而还带来了维护多种场景的负担.在使用公有云的情况下,直接依托多机房,在逻辑上提升基础服务稳定性即可.

不过确实我站的角度都是BAT/TMD这种有能力独立完整构建基础组件的企业.有大规模企业开发经验确实不会同时混用同类型存储来支撑同一份数据.因为公司内部基础框架肯定会优先支持基础场景,同时避免多套同类型轮子(虽然也不一定,头条这种就是轮子不收口的情况).当然如果公司没有这样强力的基础支撑团队,还是不建议Redis 和 memcached混用.

**Comment by draveness** [comment]
> 就算不是所有企业都是大企业,其实还是不应该在没有特殊需求的情况下混用多种同类型存储,这对系统来说并不能显著提升稳定性,反而还带来了维护多种场景的负担.在使用公有云的情况下,直接依托多机房,在逻辑上提升基础服务稳定性即可.
> 
> 不过确实我站的角度都是BAT/TMD这种有能力独立完整构建基础组件的企业.有大规模企业开发经验确实不会同时混用同类型存储来支撑同一份数据.因为公司内部基础框架肯定会优先支持基础场景,同时避免多套同类型轮子(虽然也不一定,头条这种就是轮子不收口的情况).当然如果公司没有这样强力的基础支撑团队,还是不建议Redis 和 memcached混用.

大公司其实也会走内部结算的，有的公司会签内部的 SLA。我们的业务也不信任、也不能依赖与基础服务的高可用。

说到底这东西，就是看场景

+ 不在乎可用性的就起个单机实例
+ 在乎可用性的就什么招都用上
+ 处于中间的就自己掂量，看看不同策略的 ROI

**Comment by neomerckx** [comment]
内部SLA肯定都会有,成规模的系统都需要SLA/SOP

怎么保障业务更多可能是架构上的事情,比如离线数据比对清算,比如异常重试,比如搞个错误队列delay重试.怎么在异构系统提升可用性也需要考虑业界常规实现,通用解决方案,可维护性,开发成本和开发效率.基础组件还是需要有一定的高可用性,当然在ROI可接受的程度内越高越好,这样权衡业务团队和基础开发团队的开销.

**Comment by draveness** [comment]
> 内部SLA肯定都会有,成规模的系统都需要SLA/SOP
> 
> 怎么保障业务更多可能是架构上的事情,比如离线数据比对清算,比如异常重试,比如搞个错误队列delay重试.怎么在异构系统提升可用性也需要考虑业界常规实现,通用解决方案,可维护性,开发成本和开发效率.基础组件还是需要有一定的高可用性,当然在ROI可接受的程度内越高越好,这样权衡业务团队和基础开发团队的开销.

👍 你说得对，不过这跟本文中的内容也没啥关系，也不是这里想说的问题，可以以后再讨论

**Comment by Zheaoli** [comment]
其实这里面最让人纠结的问题是什么是高可用，怎么样定义高可用

首先要明确一点的是，Infra 无论是云厂商对于用户提供的，还是内部团队使用的，都需要去保证足够的可用性，然后再谈高可用，i.e. 我们先需要保证两个9/三个9，然后再去谈其余的更高 level 的 ROI

不过我很赞同的一点是不能让业务方很傻很天真去完全相信 infra 的高可用，有句很粗暴的话叫做“你需要认为你用的东西都是一群傻逼写的”，就是要做足够的 backup 来兜底了，同时 infra 团队也许要抱有”是一群傻逼在用你的东西“的想法，去对各种奇奇怪怪的非标操作做兜底（实际上这是日常）当然这里面的平衡点就需要看具体的情况去做 trade-off 了 

当然整个业务的稳定性，是需要一个完整的体系去维护的，包括楼上说的异地多活，SOP，包括定期演练等，当然这个话题很长了。。。

**Comment by StoneKnocker** [comment]
从博主这个系列中获益良多，感谢分享。错别字勘误：”基础服务和 SLA 之前的关系“应该是”之间的关系“

---

*2020-07-10 UPDATES：已修复*

**Comment by StoneKnocker** [comment]
思考题：1.降低客户心理预期；2.如果相同服务对外承诺了不同的可用性级别，收费应该也是不一样的，那就得人为降低低收费服务的可用性，以免伤害到高付费客户

**Comment by zeer527** [comment]
文章标题欠妥

**Comment by CoreyCheung** [comment]
标题的确不合理，应该加一个修饰词，不应该过分追求高可用才对

**Comment by ichengzi** [comment]
> > 故意宕机的做法虽然由于仍然保证了SLA而不需要赔偿，但还挺狠的呀😂
> 
> 
> 
> 人家毕竟是按合同办事，而且如果真的连续几个季度都是 100% 可用，突然变成了 99.95% 可用对我们的影响更大。

可能人家也是在演练测试，注入故障，然后看系统在预计时间内能否正常恢复。🤣

**Comment by songshitong** [comment]
选择两个99.9%的厂商，同时宕机的概率是0.001*0.001=0.000001，可用的概率是99.9999%
由3个9提升到了6个9，是这样吗

**Comment by ifplusor** [comment]
不需要的只是单机无故障，而不是系统高可用
